### PR TITLE
ShaderSegements

### PR DESCRIPTION
**Pull request by petersteneteg**
OpenGL: reafactoring of the shader code parsing in ShaderOjects. Adde d support for "ShaderSegements" Small snippets that can be insertered into shader code at specific places

### CODE REVIEW

File: modules/opengl/include/modules/opengl/shader/linenumberresolver.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace inviwo {
+
+struct IVW_MODULE_OPENGL_API LineNumberResolver {
+    LineNumberResolver() = default;
+    void addLine(const std::string& file, size_t line);
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/shader/linenumberresolver.h:48]: Might want to clarify this. File is a path/key?

File: modules/opengl/include/modules/opengl/shader/linenumberresolver.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <modules/opengl/openglmoduledefine.h>
+
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace inviwo {
+
+struct IVW_MODULE_OPENGL_API LineNumberResolver {
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/shader/linenumberresolver.h:43]: Is this specific to OpenGL shaders? Perhaps we should put this in a namespace in that case?

A bit tricky to figure out what this class does just by looking at the header. What is the basis for resolving line numbers? Would be good with a  description of what this class does :)

File: modules/opengl/include/modules/opengl/shader/shadersegment.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        IVW_MODULE_OPENGL_API friend bool operator>=(const Type& lhs, const Type& rhs) {
+            return !operator<(lhs, rhs);
+        }
+
+    private:
+        std::string value_;
+    };
+
+    Type type;
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/shader/shadersegment.h]: What does type mean?

File: modules/opengl/include/modules/opengl/shader/shadersegment.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    std::string name;
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/shader/shadersegment.h:80]: name == the identifier in the shader code to replace?

File: modules/opengl/include/modules/opengl/shader/shadersegment.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    std::string snippet;
+    size_t priority = 1000;
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/shader/shadersegment.h:86]: Should mention that multiple ShaderSegments with the same identifier will be ordered according to priority. Could also mention this in the class description.

File: modules/opengl/include/modules/opengl/shader/shadersegment.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+ *********************************************************************************/
+#pragma once
+
+#include <modules/opengl/openglmoduledefine.h>
+#include <inviwo/core/common/inviwo.h>
+
+namespace inviwo {
+
+/**
+ * Hold a small text snippet to be inserted into shader code
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/shader/shadersegment.h:37]: Inserted at or replaces "name"?

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
Looks neat but some explanations should be added I think

File: modules/opengl/include/modules/opengl/shader/linenumberresolver.h (hunk 2 above)
Comment (petersteneteg) [reply modules/opengl/include/modules/opengl/shader/linenumberresolver.h:43]: This as been there for years, I just moved it into its own file

Comment (ResearchDaniel) [reply modules/opengl/include/modules/opengl/shader/linenumberresolver.h:43]: Saw that later

**Comment by petersteneteg** [comment]
@ResearchDaniel poke

File: modules/opengl/include/modules/opengl/shader/shaderobject.h [hunk 7]
```diff
@@ -132,6 +147,10 @@ class IVW_MODULE_OPENGL_API ShaderObject {
     bool hasShaderExtension(const std::string& extName) const;
     void clearShaderExtensions();
 
+    void addSegment(ShaderSegment segment);
+    void removeSegments(const std::string& key);
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/shader/shaderobject.h]: Where does the key come from? 
Might want to return the key when adding a segment?

File: modules/opengl/include/modules/opengl/shader/shadersegment.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        IVW_MODULE_OPENGL_API friend bool operator>=(const Type& lhs, const Type& rhs) {
+            return !operator<(lhs, rhs);
+        }
+
+    private:
+        std::string value_;
+    };
+
+    Type type;            //!< The placeholder that will be replaced
+    std::string name;     //!< A name of the snippet, used to the LineNumberResolver
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/shader/shadersegment.h]: Typo, used by

**Review by ResearchDaniel — APPROVED** [review]
Not sure how the shadersegment key is handled but otherwise looks good.

This will also make it easier to add other things in the future? Like special keywords (thinking of being able to derive properties based on the uniforms)?

**Comment by petersteneteg** [comment]
Not sure about adding properties from uniforms. This is mostly going the other way for adding shader code from properties. Or components, coming in a later PR...

**Pull request merged**